Add config option to leverage all cores for sabre

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

By default when running sabre in parallel we use a fixed number of threads (depending on optimization level). This was a tradeoff made for having deterministic results across multiple systems with a fixed seed set. However when running qiskit on systems with a lot of CPUs available we're leaving potential performance on the table by not using all the available cores. This new flag lets users opt-in to running sabre with n trials for n CPUs to potentially get better output results from the transpiler, with minimal to no runtime overhead, at the cost of the results not necessarily being reproducible when run on a different computer.

### Details and comments

mtreinish committed:
Add config option to leverage all cores for sabre

By default when running sabre in parallel we use a fixed number of
threads (depending on optimization level). This was a tradeoff made for
having deterministic results across multiple systems with a fixed seed
set. However when running qiskit on systems with a lot of CPUs
available we're leaving potential performance on the table by not using
all the available cores. This new flag lets users opt-in to running
sabre with n trials for n CPUs to potentially get better output results
from the transpiler, with minimal to no runtime overhead, at the cost of
the results not necessarily being reproducible when run on a different
computer.

ElePT commented on releasenotes/notes/add-sabre-all-threads-option-ad4ff7a4d045cb2b.yaml:
```diff
@@ -0,0 +1,26 @@
+
+    These transpiler passes run multiple random trials in parallel and pick
+    the result which results in the fewest :class:`.SwapGate`\s. As a rule of
+    thumb if you run more trials the provides the algorithm more opportunities
```
```suggestion
    the output which results in the fewest :class:`.SwapGate`\s. As a rule of
    thumb, if you run more trials, this provides the algorithm more opportunities
```

ElePT commented on releasenotes/notes/add-sabre-all-threads-option-ad4ff7a4d045cb2b.yaml:
```diff
@@ -0,0 +1,26 @@
+    These transpiler passes run multiple random trials in parallel and pick
+    the result which results in the fewest :class:`.SwapGate`\s. As a rule of
+    thumb if you run more trials the provides the algorithm more opportunities
+    to find a better result. By default the preset pass managers used a fixed
```
(still being used)
```suggestion
    to find a better result. By default, the preset pass managers use a fixed
```

ElePT reviewed (commented):
LGTM, not that I'll ever have the chance to see the effects of the option in my laptop 😝. I just left a few suggestions on the reno.

eliarbel commented on qiskit/transpiler/preset_passmanagers/builtin_plugins.py:
```diff
@@ -397,11 +403,16 @@ def pass_manager(self, pass_manager_config, optimization_level=None) -> PassMana
+                trial_count = CPU_COUNT
+            else:
+                trial_count = 5
+
```
It would be nice to replace this repeating block with a small util function like this:

```python
def get_trial_count(default_trials=5):
    if CONFIG.get("sabre_all_threads", None) or os.getenv("QISKIT_SABRE_ALL_THREADS"):
        return CPU_COUNT
    return default_trials
```

And use it here and in the other places below

mtreinish replied:
This is a good idea, I'll refactor this. I was also thinking about this logic overnight and was wondering if maybe it would make more sense to `max(default_trials, CPU_COUNT)` if the option is set instead of always `CPU_COUNT` . Then we don't need to caveat about not setting the flag if you have less than 20 CPUs on your local system.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/12780/commits/21ff70fc5f7c09521a3ca90a117030d2b7b7349e

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Elena Peña Tapia <57907331+ElePT@users.noreply.github.com>

mtreinish committed:
Rework logic to use the default if larger than CPU_COUNT

This commit refactors the logic added in the previous commit to a single
helper function. This reduces the code duplication and makes it easier
to work with. While doing this the logic has been updated so that when
the flag is set and the default number of trials is larger than the
CPU_COUNT we use the default. This means the logic when the flag is set
is to run `max(default_trials, CPU_COUNT)` which should better match
user expectations around the flag.

eliarbel reviewed (approved):
LGTM

ElePT reviewed (approved):
LGTM too
